src/common/trinamic.cpp

```diff
@@ -35,51 +32,52 @@ void tmc_delay(uint16_t time) // delay for switching tmc step pin level
 
 void init_tmc(void) {
 
-    pStepX = &stepperX;
-    pStepY = &stepperY;
-    pStepZ = &stepperZ;
-    pStepE = &stepperE0;
-    //int i = 0;
-    pStepX->TCOOLTHRS(400);
-    pStepY->TCOOLTHRS(400);
-    pStepZ->TCOOLTHRS(400);
-    pStepE->TCOOLTHRS(400); //400
-    pStepX->SGTHRS(140);
-    pStepY->SGTHRS(130);
-    pStepZ->SGTHRS(100);
-    pStepE->SGTHRS(100);
+    pStep[0] = &stepperX;
+    pStep[1] = &stepperY;
+    pStep[2] = &stepperZ;
+    pStep[3] = &stepperE0;
+    pStep[0]->TCOOLTHRS(400);
+    pStep[1]->TCOOLTHRS(400);
+    pStep[2]->TCOOLTHRS(400);
+    pStep[3]->TCOOLTHRS(400);
+    pStep[0]->SGTHRS(140);
+    pStep[1]->SGTHRS(130);
+    pStep[2]->SGTHRS(100);
+    pStep[3]->SGTHRS(100);
 }
 
-void tmc_sample(void) {
-    unsigned int sgx = pStepX->SG_RESULT();
-    //unsigned int sgy = pStepY->SG_RESULT();
-    unsigned int diag = 0;
-    //	diag |= hwio_di_get_val(_DI_Z_DIAG);
-    //diag |= hwio_di_get_val(_DI_Y_DIAG) << 1;
-    //diag |= hwio_di_get_val(_DI_Z_DIAG) << 2;
-    //diag |= hwio_di_get_val(_DI_E0_DIAG) << 3;
-    unsigned int tstepx = pStepX->TSTEP();
-    //unsigned int tstepy = pStepX->TSTEP();
-    //	int sgz = pStepZ->SG_RESULT();
-    //	int sge = pStepE->SG_RESULT();
-    //	_dbg("sg %d %d %d %d", sgx, sgy, sgz, sge);
-    sgx = sgx;       //prevent warning
-    tstepx = tstepx; //prevent warning
-    diag = diag;     //prevent warning
-    DBG("sg %u %u %u", sgx, diag, tstepx);
+// this function performs stallguard sample for single axis
+// return value contain bitmask of sampled axis, in case of nonzero return value this call take 5ms
+//  Now we are using stepper.axis_is_moving and value is set to zero for stopped motor,
+//  right way is reading tstep and comparing it to TCOOLTHRS, but it takes time (read register).
+//  Using stepper.axis_is_moving is simple but in some cases we get bad samples (tstep > TCOOLTHRS).
+//  Maybe we can improve this by calculating tstep from stepper variables.
+uint8_t tmc_sample(void) {
+    uint8_t mask = 0;
+    if (tmc_sg_mask) {
+        while ((tmc_sg_mask & (1 << tmc_sg_axis)) == 0)
+            tmc_sg_axis = (tmc_sg_axis + 1) & 0x03;
+        if (stepper.axis_is_moving((AxisEnum)tmc_sg_axis)) {
+            mask = (1 << tmc_sg_axis);
+            tmc_sg[tmc_sg_axis] = pStep[tmc_sg_axis]->SG_RESULT();
+        } else
+            tmc_sg[tmc_sg_axis] = 0;
+        tmc_sg_axis = (tmc_sg_axis + 1) & 0x03;
+    }
+    return mask;
 }
 
 void tmc_set_sgthrs(uint16_t SGT) {
-    pStepX->SGTHRS(SGT);
-    pStepY->SGTHRS(SGT);
-    pStepZ->SGTHRS(SGT);
-    pStepE->SGTHRS(SGT);
+    pStep[0]->SGTHRS(SGT);
```

Comment:
can we have symbolic constant names for the indices in pStep[]?
like `enum class Axes : uint_least8_t { X = 0, Y, Z, E }` ?

Reply:
commit 45ddd2c